File: example/android/app/src/main/java/preeternal/reactnativefilehash/example/BenchmarkFileModule.kt
package preeternal.reactnativefilehash.example

import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.io.File
import java.io.FileOutputStream

class BenchmarkFileModule(
  private val reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String = NAME

  @ReactMethod
  fun createFile(sizeBytes: Double, promise: Promise) {
    try {
      val requestedSize = sizeBytes.toLong()
      if (requestedSize <= 0L) {
        promise.reject("E_INVALID_SIZE", "Benchmark file size must be positive")
        return
      }

      val directory = File(reactContext.cacheDir, "zfh-benchmark")
      if (!directory.exists() && !directory.mkdirs()) {
        promise.reject("E_CREATE_DIR", "Failed to create benchmark cache directory")
        return
      }

      val file = File(directory, "payload-$requestedSize.bin")
      if (file.exists() && file.length() == requestedSize) {
        promise.resolve(file.absolutePath)
        return
      }

      if (file.exists() && !file.delete()) {
        promise.reject("E_DELETE_FILE", "Failed to replace benchmark file")
        return
      }

      writeDeterministicFile(file, requestedSize)
      promise.resolve(file.absolutePath)
    } catch (error: Exception) {
      promise.reject("E_CREATE_FILE", "Failed to create benchmark file", error)
    }
  }

  @ReactMethod
  fun log(message: String) {
    Log.i("ZFHBenchmark", message)
  }

  private fun writeDeterministicFile(file: File, sizeBytes: Long) {
    val buffer = ByteArray(CHUNK_SIZE)
    for (i in buffer.indices) {
      buffer[i] = (i and 0xff).toByte()
    }

    FileOutputStream(file, false).use { output ->
      var remaining = sizeBytes
      while (remaining > 0L) {
        val count = minOf(buffer.size.toLong(), remaining).toInt()
        output.write(buffer, 0, count)
        remaining -= count.toLong()
      }
      output.fd.sync()
    }
  }

  companion object {
    const val NAME = "BenchmarkFile"
    private const val CHUNK_SIZE = 1024 * 1024
  }
}

File: example/ios/FileHashExample/BenchmarkFileModule.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

static const NSUInteger ZFHBenchmarkChunkSize = 1024 * 1024;

@interface BenchmarkFileModule : NSObject <RCTBridgeModule>
@end

@implementation BenchmarkFileModule

RCT_EXPORT_MODULE(BenchmarkFile)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

RCT_REMAP_METHOD(createFile,
                 createFileWithSizeBytes:(nonnull NSNumber *)sizeBytes
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^{
    long long requestedSize = sizeBytes.longLongValue;
    if (requestedSize <= 0) {
      reject(@"E_INVALID_SIZE", @"Benchmark file size must be positive", nil);
      return;
    }

    NSFileManager *fileManager = NSFileManager.defaultManager;
    NSURL *cacheURL = [fileManager URLsForDirectory:NSCachesDirectory
                                          inDomains:NSUserDomainMask].firstObject;
    NSURL *directoryURL = [cacheURL URLByAppendingPathComponent:@"zfh-benchmark"
                                                    isDirectory:YES];

    NSError *error = nil;
    if (![fileManager createDirectoryAtURL:directoryURL
               withIntermediateDirectories:YES
                                attributes:nil
                                     error:&error]) {
      reject(@"E_CREATE_DIR", @"Failed to create benchmark cache directory", error);
      return;
    }

    NSString *fileName = [NSString stringWithFormat:@"payload-%lld.bin", requestedSize];
    NSURL *fileURL = [directoryURL URLByAppendingPathComponent:fileName];
    NSDictionary<NSFileAttributeKey, id> *attributes =
        [fileManager attributesOfItemAtPath:fileURL.path error:nil];
    if ([attributes[NSFileSize] unsignedLongLongValue] == (unsigned long long)requestedSize) {
      resolve(fileURL.path);
      return;
    }

    if ([fileManager fileExistsAtPath:fileURL.path] &&
        ![fileManager removeItemAtURL:fileURL error:&error]) {
      reject(@"E_DELETE_FILE", @"Failed to replace benchmark file", error);
      return;
    }

    if (![fileManager createFileAtPath:fileURL.path contents:nil attributes:nil]) {
      reject(@"E_CREATE_FILE", @"Failed to create benchmark file", nil);
      return;
    }

    NSFileHandle *handle = [NSFileHandle fileHandleForWritingToURL:fileURL error:&error];
    if (handle == nil) {
      reject(@"E_OPEN_FILE", @"Failed to open benchmark file", error);
      return;
    }

    NSMutableData *chunk = [NSMutableData dataWithLength:ZFHBenchmarkChunkSize];
    uint8_t *bytes = chunk.mutableBytes;
    for (NSUInteger i = 0; i < ZFHBenchmarkChunkSize; i += 1) {
      bytes[i] = (uint8_t)(i & 0xff);
    }

    long long remaining = requestedSize;
    while (remaining > 0) {
      @autoreleasepool {
        NSUInteger count =
            (NSUInteger)((remaining < (long long)ZFHBenchmarkChunkSize)
                             ? remaining
                             : (long long)ZFHBenchmarkChunkSize);
        if (count == ZFHBenchmarkChunkSize) {
          [handle writeData:chunk];
        } else {
          [handle writeData:[chunk subdataWithRange:NSMakeRange(0, count)]];
        }
        remaining -= (long long)count;
      }
    }

    [handle synchronizeFile];
    [handle closeFile];
    resolve(fileURL.path);
  });
}

RCT_EXPORT_METHOD(log:(NSString *)message)
{
  NSLog(@"%@", message);
}

@end
